JOSS paper 2024 (#491)
trixi-framework/TrixiParticles.jl: paper/paper.md

* *Structure Systems*
    + `TotalLagrangianSPHSystem` (TLSPH): System to simulate elastic structure where the interaction is defined with stress tensors [@O_Connor:2021].
    + `DEMSystem`: Discrete element system [@Bicanic:2004], [@Cundall:1979].

* *Boundary Systems*
    + `BoundarySPHSystem` with several boundary models where each model follows a different interaction rule.
    + `OpenBoundarySPHSystem`: System to simulate non-reflecting (open) boundary conditions [@Lastiwka:2009]

* *Correction methods and models*
  + Density diffusion [@Antuono:2010]
  + Transport-velocity formulation (TVF) [@Adami:2013]
  + Intra-particle-force surface tension [@Akinci:2013]
  + Kernel corrections TODO: refs

* *TODO: category naming?*
  + Efficient grid neighborhood search
  + GPU support

\autoref{fig:falling_sphere} illustrates an example of our simulation results. In this example, an elastic sphere modeled with TLSPH falls into a tank filled with water, modeled by WCSPH.

![Elastic sphere falling into a tank filled with water. \label{fig:falling_sphere}](falling_sphere_combined_nonstick_4k_178.png){width=100%}

> The current state allows also to validate our simulation and produce quantitative results with a post-process callback. \autoref{fig:beam_y_deflection} shows simulation results of TrixiParticles.jl (on the left) and [@O_Connor:2021] (on the right) compared against a reference value of [@Turek:2007].

Reviewer: ```suggestion
Validation can be performed by, for example, quantitatively comparing results using a post-process callback. \autoref{fig:beam_y_deflection} shows simulation results of TrixiParticles.jl (on the left) and [@O_Connor:2021] (on the right) compared against a reference value of [@Turek:2007].
```

The curves show the y-deflection of the tip of a beam oscillating under its own weight.
The results obtained with TrixiParticles.jl perfectly match those of [@O_Connor:2021].

![Comparison of TrixiParticles.jl and  [@O_Connor:2021] against [@Turek:2007]: Tip y-deflection of an oscillating beam with different resolutions, where $t_s$ is the thickness of the beam and $dp$ is the particle spacing. \label{fig:beam_y_deflection}](oscillating_beam.png){width=100%}

> As a young project that aims to be the Julia code for particle-based simulation we expect to integrate other methods such as incompressible SPH (ISPH) [@Liu:2022] or heat conduction with SPH [@Biriukov:2018] and couple them among each other and even with mesh-based methods.

Reviewer: ```suggestion
As a young project that aims to be the Julia code for particle-based simulation we expect to integrate other methods such as incompressible SPH (ISPH) [@Liu:2022] or heat conduction with SPH [@Biriukov:2018], and DEM specifically tailored for applications in coatings [@Ghasemi:2019] [@Laukannen:2020]. 
```
https://doi.org/10.1016/j.surfcoat.2018.11.068
https://doi.org/10.1016/j.triboint.2020.106534

Reviewer: Forgot the second part:
In the future these methods will be coupled with each other and even with mesh-based methods to provide a comprehensive framework.

---

# Summary

TrixiParticles.jl, part of the Trixi Framework [@schlottkelakemper2020trixi], is an innovative Julia-based open-source software designed for particle-based multiphysics simulations.
In this sense, we distinguish two approaches: Particles are either treated as physical particles or as mathematical interpolation points.
The former case refers to methods that model separate, discrete particles with rotational degrees of freedom such as the Discrete Element Method (DEM) proposed by [@Cundall:1979].
Whereas the latter case refers to the smoothed particle hydrodynamics (SPH) which is a numerical discretization method for solving continuum mechanic problems.
SPH was originally developed by [@Monaghan:1977] to simulate astrophysical applications and is now widely used to simulate computational fluid dynamics, structural mechanics and even heat conduction problems.

Here, we give a brief overview of the software package TrixiParticles.jl, starting with a description of the scientific background before going on to describe the functionality and benefit in more detail.
Finally, a few exemplary results and implemented features are briefly presented.

# Statement of need

> Numerical simulations such as computational fluid dynamics, structural mechanics, thermodynamics, or magnetohydrodynamics present several challenges especially when simulating real-world problems.

Reviewer: ```suggestion
Numerical simulations, such as computational fluid dynamics, structural mechanics, thermodynamics, or magnetohydrodynamics pose several challenges, especially when simulating real-world problems.
```

> Simulating these problems involves challenges such as dealing with complex geometries, free surfaces,

Reviewer: ```suggestion
These simulations involve challenges such as dealing with complex geometries, free surfaces,
```

> deformable boundaries or moving material interfaces and coupling of different physics disciplines.

Reviewer: ```suggestion
deformable boundaries and moving material interfaces as well as the coupling of different physics disciplines.
```

> The lagrangian formalism in particle-based methods allows particles to move along a velocity field without any connection to neighboring particles and, thus, eliminating the need for a mesh to discretize the simulation domain.

Reviewer: ```suggestion
The Lagrangian formalism in particle-based methods allows particles to move along a velocity field without any connection to neighboring particles, thus eliminating the need for a mesh to discretize the simulation domain.
```

> The mesh-free formalism not only allows for easy preprocessing, making them particularly suitable for simulating complex geometries, but also allows for the straightforward handling of very large deformations.

Reviewer: ```suggestion
This mesh-free approach not only simplifies the preprocessing, making it particularly suitable for simulating complex geometries, but it also facilitates simulations of large deformations and movements.
```

In addition, particle-based methods are inherently suited to simulating free surfaces, material interfaces and challenges involving moving boundaries.

> Facilitating the transition from continuum to fragmentation becomes straightforward by representing each material with its own set of particles, thereby effectively addressing material interface issues.

Reviewer: ```suggestion
Facilitating the transition from continuum to fragmentation becomes straightforward by representing each material with its own set of particles, effectively addressing material interface issues.
```

This also simplifies the coupling of different singlephysics into a multiphysics system.

> There are several open-source software projects specialized for SPH methods, including DualSPHysics [@Dominguez:2021], SPlisHSPlasH [@Bender] and SPHinXsys [@Zhang:2021],

Reviewer: ```suggestion
Several open-source software projects specialize in SPH methods, including DualSPHysics [@Dominguez:2021], SPlisHSPlasH [@Bender] and SPHinXsys [@Zhang:2021],
```

written in C++, and  PySPH [@Ramachandran:2021], written in Python.
These softwares reflect the advantages of the SPH methods by simulating problems such as fluid-structure interaction (FSI) and free surfaces [@O_Connor:2021] or complex geometries [@Laha:2024].

TrixiParticles.jl is written in Julia and thus combines the advantage of easy and rapid prototyping with the ability of high performance computing using multicore parallelization and hardware accelerators.
The software package provides support for developing or testing new SPH methods and also for simulating or coupling other particle-based methods such as DEM.
Another aspect is the ability to add custom methods or particle interactions without modifying the source code, since each simulation is configured and set up by pure Julia code.

# Overview of particle-based simulation

> In TrixiParticles.jl, particles of a single particle-based method are grouped into a so called \emph{system}.

Reviewer: ```suggestion
In TrixiParticles.jl, particles of a single particle-based method are grouped into what we refer to as a \emph{system}.
```

> The interaction between two particles is defined by the types of their systems. This approach makes it easy to add new methods and different physics.

Reviewer: ```suggestion
The interactions between two particles are determined by the system types to which they belong. This approach makes it easy to add new methods and physical models.
```

![Particles of two different systems in a simulation domain. The black and gray dashed circle represent the search radius for neighboring particles of particle $a$ and $b$, respectively.\label{fig:systems}](systems.png){width=40%}

> To illustrate this, \autoref{fig:systems} shows particles in a simulation domain. The black particles belong to system $\mathcal{S}_1$ and the gray particles belong to system $\mathcal{S}_2$.

Reviewer: ```suggestion
To illustrate this, \autoref{fig:systems} depicts particles within a simulation domain. The black particles belong to system $\mathcal{S}_1$ and the gray particles belong to system $\mathcal{S}_2$.
```

In general, the force $f_a$ experienced by a particle $a$ is calculated as
$$ f_a = \sum_{b \in \mathcal{S}_1} f_{ab}^{\mathcal{S}_1} + \sum_{b \in \mathcal{S}_2} f_{ab}^{\mathcal{S}_2} + \dots + \sum_{b\in \mathcal{S}_n}f_{ab}^{\mathcal{S}_n}, $$
where the interaction force $f_{ab}^{\mathcal{S}_i}$ that particle $a$ experiences due to particle $b$ depends on the system type of particle $a$, the system type $\mathcal{S}_i$ of particle $b$ and the relative particle distance.
For computational efficiency, only particles with a distance within a system-dependent search radius interact with each other.

> For example, the SPH method determines the force between two SPH particles according to [@Monaghan:2005] as

Reviewer: ```suggestion
For example, in the SPH method the force between two particles is determined according to [@Monaghan:2005] as
```

$$ f_{ab} = -m_a m_b \left( \frac{p_a}{\rho_a^2} + \frac{p_b}{\rho_b^2} \right) \nabla_a W_{ab} + \Pi_{ab},$$
where $m_a$, $m_b$, $\rho_a$, $\rho_b$, $p_a$, $p_b$ are the mass, density and pressure of particles $a$ and $b$, respectively. The last term $\Pi_{ab}$ includes dissipative terms such as artificial viscosity [@Monaghan:2005] and is scheme-specific. The weighting function $W_{ab}$, also called kernel-function, depends on the relative distance between particles $a$ and $b$.

# Code structure and features

> \autoref{fig:structure} depicts the basic building blocks of TrixiParticles.jl. A simulation essentially consists of spatial discretization (left block) and time integration (center block). For the latter, the Julia package [OrdinaryDiffEq.jl](https://github.com/SciML/OrdinaryDiffEq.jl) is used. The callbacks (right block) provide additional functionality and communicate with the time integration method during the simulation.

Reviewer: ```suggestion
\autoref{fig:structure} illustrates the fundamental components of TrixiParticles.jl. A simulation essentially consists of spatial discretization (left block) and time integration (center block). For the latter, the Julia package [OrdinaryDiffEq.jl](https://github.com/SciML/OrdinaryDiffEq.jl) is used. The callbacks (right block) provide additional functionality and communicate with the time integration method during the simulation.
```

The semidiscretization couples the systems of a simulation and also creates the corresponding neighborhood searches for each system. The resulting ordinary differential equation (ODE) problem is then fed into the time integrator and is solved by an appropriate numerical time integration scheme.

![Main building blocks of TrixiParticles.jl \label{fig:structure}](structure.png){width=100%}

> TrixiParticles.jl is open source and available under the MIT license at [GitHub](https://github.com/trixi-framework/TrixiParticles.jl),  along with detailed [documentation](https://trixi-framework.github.io/TrixiParticles.jl/stable/) explaining how to use it. Additionally, we provide tutorials explaining how to set up a simulation of fluidflows, structure mechanics, or fluid-structure interaction.

Reviewer: ```suggestion
TrixiParticles.jl is open source and available under the MIT license at [GitHub](https://github.com/trixi-framework/TrixiParticles.jl),  along with detailed [documentation](https://trixi-framework.github.io/TrixiParticles.jl/stable/) explaining how to use it. Additionally, we provide tutorials explaining how to set up a simulation for a wide range of topics including fluid flows, structure mechanics, or fluid-structure interaction.
```

> A collection of simulation setups to get started with can be found in the `examples` directory.

Reviewer: ```suggestion
Users can find a variety of simulation setups in the examples directory to help get started.
```

### Feature highlights

As a young project that aims to be the Julia code for particle-based simulation we expect to integrate other methods such as incompressible SPH (ISPH) [@Liu:2022] or heat conduction with SPH [@Biriukov:2018] and couple them among each other and even with mesh-based methods.

So far, the following feature highlights have been implemented:

* Weakly compressible SPH with density diffusion [@Antuono:2010]

* Entropically damped artificial compressibility (EDAC) [@Ramachandran:2019]

* Total Lagrangian SPH (TLSPH) and fluid-structure interaction [@O_Connor:2021]

* Discrete element method (DEM) [@Bicanic:2004], [@Cundall:1979]

* Transport-velocity formulation (TVF) [@Adami:2013]

* Intra-particle-force surface tension [@Akinci:2013]

* Non-reflecting (open) boundaries [@Lastiwka:2009]

* Efficient grid neighborhood search

* GPU support

Figure \autoref{fig:falling_sphere} illustrates an example of our simulation results. In this example, an elastic sphere modeled with TLSPH falls into a tank filled with water.

> ![Elastic sphere falling into a tank filled with water. \label{fig:falling_sphere}](falling_sphere_combined_nonstick_4k_178.png){width=50%}

Author: uploaded where? No, this is the file you sent me on slack.

> ![Elastic sphere falling into a tank filled with water. \label{fig:falling_sphere}](falling_sphere_combined_nonstick_4k_178.png){width=50%}

Author: Yes, 16:9 would look better

    orcid: 0000-0001-9788-5949
    equal-contrib: true # (This is how you can denote equal contributions between multiple authors)
    affiliation: 2
  - name: Sven Berger
    orcid: 0000-0001-6083-7038
    equal-contrib: true # (This is how you can denote equal contributions between multiple authors)
    affiliation: 3
  - name: Michael Schlottke-Lakemper
    orcid: 0000-0002-3195-2536
    affiliation: 1 # TODO: @Michael: Stuttgart or Augsburg
  - name: Gregor J. Gassner
    orcid: 0000-0002-1752-1158
    affiliation: 2
affiliations:
 - name: High-Performance Computing Center Stuttgart, University of Stuttgart, Germany
   index: 1
 - name: Department of Mathematics and Computer Science, University of Cologne, Germany
   index: 2
 - name: Institute of Surface Science, Helmholtz-Zentrum hereon, Germany
   index: 3
date: 27 March 2024 # TODO
bibliography: paper.bib

# Optional fields if submitting to a AAS journal too, see this blog post:
# https://blog.joss.theoj.org/2018/12/a-new-collaboration-with-aas-publishing
aas-doi: 10.3847/xxxxx <- update this with the DOI from AAS once you know it.
aas-journal: Astrophysical Journal <- The name of the AAS journal.
---

# Summary

> TrixiParticles.jl, part of the Trixi Project[@trixi], is an innovative Julia-based open-source framework designed for  particle-based multiphysics simulations. It aims to overcome the limitations of traditional mesh-based methods in handling complex geometries and specialized applications, such as computational fluid dynamics and structural mechanics, by offering a versatile platform for Smoothed Particle Hydrodynamics (SPH) and the Discrete Element Method (DEM), among others. Unique in its approach, TrixiParticles.jl facilitates the easy addition of new particle systems and their interactions, enhancing research flexibility. This is further complemented by its user-friendly design, which allows simulations to be configured directly with Julia code, eliminating the need for parameter files. This feature not only simplifies the integration of custom functionalities but also promotes rapid prototyping and efficient scaling of simulations, establishing TrixiParticles.jl as a robust and adaptable tool for advancing multiphysics simulations in various scientific and engineering fields.

Author: > I think it would make review easier if you could put each sentence in a separate line of code.

Good point. I didn't notice that since I have automatic word wrap enabled for markdowns.

Finally, some of the results obtained and some of the features that have been implemented so far are briefly presented.

# Statement of need

Numerical simulations such as computational fluid dynamics, structural mechanics, thermodynamics
or magnetohydrodynamics are often simulated using mesh-based methods. A wide range of
mature and powerful software is available for these simulations (TODO: cite OpenFoam etc?). However, in certain applications,
these methods reach their limits or are not designed to solve these problems efficiently (TODO: Ref to biomechanic, welding processes etc? ).
This is where particle-based methods offer an alternative.

The mesh-free formalism of the methods allows for easy preprocessing, making them particularly
suitable for simulating complex geometries. This also facilitates the coupling of different single
physics into a multiphysics system.

There are several open-source software e.g. specialized for SPH methods, including [@DualSPHysics], [@SplisHSPlasH] and [@SPHinXsys],
written in C++, and  [@PySPH], written in Python. These frameworks emphasize performance and are designed for a variety of SPH schemes.

> TrixiParticles.jl provides support not only for SPH methods but also for simulating or coupling other particle-based methods such as DEM.

Author: > TrixiParticles.jl provides support for developing or testing new SPH methods and also for simulating or coupling other particle-based methods such as DEM.

Is this better? Sounds a bit weird

Another aspect is that TrixiParticles.jl is not configured at runtime via a parameter file, as it is the case with other numerical simulation codes.
Instead, each simulation is configured and set up by pure Julia code.
This makes it easy for users to add custom functionality even without touching the source code.

> TODO ?: As a side note, the existing code also supports GPU implementation without the need for separate code.

Reviewer: I have already proposed how that could be included in the text.

# Functionality

\autoref{fig:structure} depicts the basic building blocks of TrixiParticles.jl. A simulation essentially consists of spatial discretization (left block) and time integration (center block). For the latter, the [SCiML ecosystem for ordinary differential equations](https://docs.sciml.ai/DiffEqDocs/latest/) (ODE) is used. The callbacks (right block) are used to communicate with the time integration interface during the simulation.

The object [`Semidiscretization`](https://trixi-framework.github.io/TrixiParticles.jl/stable/general/semidiscretization/) couples the passed systems to one simulation and also creates the corresponding neighborhood searches for each system. The resulting ODE problem is then fed into the time integrator and is solved by an appropriate numerical time integration scheme.

> ![TODO: caption *Inspired by [docs Trixi](https://trixi-framework.github.io/Trixi.jl/stable/overview/#overview-semidiscretizations)* \label{fig:structure}](structure.png){width=100%}

Reviewer: Finish the diagram please.

* *Boundary Systems*
    + [`BoundarySPHSystem`](https://trixi-framework.github.io/TrixiParticles.jl/stable/systems/boundary/) with several boundary models where each model follows a different interaction rule.
    + `OpenBoundarySPHSystem`: System to simulate non-reflecting boundary conditions [@Lastiwka:2009]

The code presented here can be found on [GitHub](https://github.com/trixi-framework/TrixiParticles.jl),  along with a detailed [manual](https://trixi-framework.github.io/TrixiParticles.jl/stable/) explaining how to use the package. Additionally, we provide tutorials explaining how to setup a simulation for fluids, structure mechanics or a fluid-structure interaction. All implemented examples for testing are located in the [`examples/`](https://github.com/trixi-framework/TrixiParticles.jl/tree/main/examples) folder, thus, the results can be easily reproduced.

### Feature highlights

As a young project that aims to be the Julia code for particle-based simulation we expect to integrate other methods such as incompressible SPH (ISPH) [@Liu:2022] or heat conduction with SPH [@Biriukov:2018] and couple them among each other and even with mesh-based methods.

> TODO: Instead of listing it in the paper, it might be better to link the [feature list](https://trixi-framework.github.io/TrixiParticles.jl/stable/#Features)

Reviewer: That is not a good idea too only link there since the URL might change in the future.

  - name: Michael Schlottke-Lakemper
    orcid: 0000-0002-3195-2536
    affiliation: 1, 4
affiliations:
 - name: High-Performance Computing Center Stuttgart, University of Stuttgart, Germany
   index: 1
 - name: Department of Mathematics and Computer Science, University of Cologne, Germany
   index: 2
 - name: Institute of Surface Science, Helmholtz-Zentrum hereon, Germany
   index: 3
 - name: High-Performance Scientific Computing, University of Augsburg, Germany
   index: 4
date: \today
bibliography: paper.bib
---

# Summary

TrixiParticles.jl, part of the Trixi Framework [@schlottkelakemper2020trixi], is an innovative Julia-based open-source software designed for particle-based multiphysics simulations.
In this sense, we distinguish two approaches: Particles are either treated as physical particles or as mathematical interpolation points.
The former case refers to methods that model separate, discrete particles with rotational degrees of freedom such as the Discrete Element Method (DEM) proposed by [@Cundall:1979].
Whereas the latter case refers to the smoothed particle hydrodynamics (SPH) which is a numerical discretization method for solving continuum mechanic problems.

> SPH was originally developed by [@Monaghan:1977] to simulate astrophysical applications and is now widely used to simulate computational fluid dynamics, structural mechanics and even heat conduction problems.

Reviewer 2: Does this fit in a summary? I agree that the previous version was a bit too much advertisement, but explaining that particles can either be physical particles or interpolation points seems a bit misplaced to me in the summary.

Finally, a few exemplary results and implemented features are briefly presented.

# Statement of need

Numerical simulations, such as computational fluid dynamics, structural mechanics, thermodynamics, or magnetohydrodynamics pose several challenges, especially when simulating real-world problems.
These simulations involve challenges such as dealing with complex geometries, free surfaces,
deformable boundaries and moving material interfaces as well as the coupling of different physics disciplines.

The Lagrangian formalism in particle-based methods allows particles to move along a velocity field without any connection to neighboring particles, thus eliminating the need for a mesh to discretize the simulation domain.
This mesh-free approach not only simplifies the preprocessing, making it particularly suitable for simulating complex geometries, but it also facilitates simulations of large deformations and movements.
In addition, particle-based methods are inherently suited to simulating free surfaces, material interfaces and challenges involving moving boundaries.

> Facilitating the transition from continuum to fragmentation becomes straightforward by representing each material with its own set of particles, effectively addressing material interface issues.

Reviewer 2: I don't understand this sentence.

Author: Which part of it?

Reviewer 2: "transition from continuum to fragmentation" and "material interface issues"

Author: > "transition from continuum to fragmentation"

In contrast to mesh-based methods, it's is no problem when, e.g., parts of the fluid detach from the "entire" fluid. That is, single droplets.

> material interface issues

Interface problems are trivial with particle-based methods. That is, simulations with more than one material. 
But more challenging with mesh-based methods.

Reviewer 2: I still find it hard to understand. If you want to keep it, could you at least replace "facilitating"? IMO, that doesn't fit at all. "Making the transition easier becomes straightforward"?

> This also simplifies the coupling of different singlephysics into a multiphysics system.

Reviewer 2: singlephysics? Maybe single-physics?
```suggestion
This also simplifies the coupling of different single-physics into a multi-physics system.
```

Author: I struggled too. But "multiphysics" is also written without a hyphen (e.g https://en.wikipedia.org/wiki/Multiphysics_simulation)

Reviewer 3: I don't think "singlephysics" is ever used. I'd write something like
> [...] coupling of different physical systems into a multi-physics setup.

There are several open-source software projects specialized for SPH methods, including DualSPHysics [@Dominguez:2021], SPlisHSPlasH [@Bender] and SPHinXsys [@Zhang:2021],
written in C++, and  PySPH [@Ramachandran:2021], written in Python.
These softwares reflect the advantages of the SPH methods by simulating problems such as fluid-structure interaction (FSI) and free surfaces [@O_Connor:2021] or complex geometries [@Laha:2024].

TrixiParticles.jl is written in Julia and thus combines the advantage of easy and rapid prototyping with the ability of high performance computing using multicore parallelization and hardware accelerators.

> The software package provides support for developing or testing new SPH methods and also for simulating or coupling other particle-based methods such as DEM.

Reviewer 2: ```suggestion
The software package provides support for developing and testing new SPH methods and also for simulating and coupling other particle-based methods such as DEM.
```

Another aspect is the ability to add custom methods or particle interactions without modifying the source code, since each simulation is configured and set up by pure Julia code.

# Overview of particle-based simulation

In TrixiParticles.jl, particles of a single particle-based method are grouped into what we refer to as a \emph{system}.
The interaction between two particles is defined by the types of their systems. This approach makes it easy to add new methods and different physics.

> ![Particles of two different systems in a simulation domain. The black and gray dashed circle represent the search radius for neighboring particles of particle $a$ and $b$, respectively.\label{fig:systems}](systems.png){width=40%}

Reviewer 2: ```suggestion
![Particles of two different systems in a simulation domain. The black and gray dashed circles represent the search radii for neighboring particles of particles $a$ and $b$, respectively.\label{fig:systems}](systems.png){width=40%}
```
Also, can you please change the symbols in the image accordingly to `a` and `b` without the superscripts?

To illustrate this, \autoref{fig:systems} depicts particles within a simulation domain. The black particles belong to system $\mathcal{S}_1$ and the gray particles belong to system $\mathcal{S}_2$.
In general, the force $f_a$ experienced by a particle $a$ is calculated as
$$ f_a = \sum_{b \in \mathcal{S}_1} f_{ab}^{\mathcal{S}_1} + \sum_{b \in \mathcal{S}_2} f_{ab}^{\mathcal{S}_2} + \dots + \sum_{b\in \mathcal{S}_n}f_{ab}^{\mathcal{S}_n}, $$

> where the interaction force $f_{ab}^{\mathcal{S}_i}$ that particle $a$ experiences due to particle $b$ depends on the system type of particle $a$, the system type $\mathcal{S}_i$ of particle $b$ and the relative particle distance.

Reviewer 2: ```suggestion
where the interaction force $f_{ab}^{\mathcal{S}_i}$ that particle $a$ experiences due to particle $b$ depends on the system type of particle $a$, the system type $\mathcal{S}_i$ of particle $b$, and the relative particle distance.
```

For computational efficiency, only particles with a distance within a system-dependent search radius interact with each other.

For example, the SPH method determines the force between two SPH particles according to [@Monaghan:2005] as
$$ f_{ab} = -m_a m_b \left( \frac{p_a}{\rho_a^2} + \frac{p_b}{\rho_b^2} \right) \nabla_a W_{ab} + \Pi_{ab},$$
where $m_a$, $m_b$, $\rho_a$, $\rho_b$, $p_a$, $p_b$ are the mass, density and pressure of particles $a$ and $b$, respectively. The last term $\Pi_{ab}$ includes dissipative terms such as artificial viscosity [@Monaghan:2005] and is scheme-specific. The weighting function $W_{ab}$, also called kernel-function, depends on the relative distance between particles $a$ and $b$.

# Code structure and features

\autoref{fig:structure} depicts the basic building blocks of TrixiParticles.jl. A simulation essentially consists of spatial discretization (left block) and time integration (center block). For the latter, the Julia package [OrdinaryDiffEq.jl](https://github.com/SciML/OrdinaryDiffEq.jl) is used. The callbacks (right block) provide additional functionality and communicate with the time integration method during the simulation.

> The semidiscretization couples the systems of a simulation and also creates the corresponding neighborhood searches for each system. The resulting ordinary differential equation (ODE) problem is then fed into the time integrator and is solved by an appropriate numerical time integration scheme.

Reviewer 2: ```suggestion
The semidiscretization couples the systems of a simulation and also manages the corresponding neighborhood searches for each system. The resulting ordinary differential equation (ODE) problem is then fed into the time integrator and is solved by an appropriate numerical time integration scheme.
```
It might not create them in the future.

![Main building blocks of TrixiParticles.jl \label{fig:structure}](structure.png){width=100%}

TrixiParticles.jl is open source and available under the MIT license at [GitHub](https://github.com/trixi-framework/TrixiParticles.jl),  along with detailed [documentation](https://trixi-framework.github.io/TrixiParticles.jl/stable/) explaining how to use it. Additionally, we provide tutorials explaining how to set up a simulation of fluid flows, structure mechanics, or fluid-structure interaction.
A collection of simulation setups to get started with can be found in the examples directory.

## Feature highlights

So far, the following feature highlights have been implemented:

* *Fluid Systems*
    + Weakly compressible SPH (WCSPH): Standard SPH method originally developed by [@Monaghan:1977] to simulate astrophysics applications.

>     + Entropically damped artificial compressible (EDAC) SPH: As opposed to the WCSPH scheme, which uses an equation of state, this scheme uses a pressure evolution equation to calculate the pressure, which is derived by [@Clausen:2013] and adapted to SPH by [@Ramachandran:2019].

Reviewer 2: ```suggestion
    + Entropically damped artificially compressible (EDAC) SPH: As opposed to the WCSPH scheme, which uses an equation of state, this scheme uses a pressure evolution equation to calculate the pressure, which is derived by [@Clausen:2013] and adapted to SPH by [@Ramachandran:2019].
```

Reviewer 2: Is the SPH artificial or the compressibility?
The original name is "entropically damped artificial compressibility", isn't it? So that makes the compressibility artificial, which makes it "artificially compressible".

Author: In general, the WCSPH is an artificial compressibility (AC) scheme. 
According to [Clausen 2013](http://dx.doi.org/10.1103/PhysRevE.87.013309):
"Traditional AC schemes rely on an artificial equation of state that gives the pressure as a function of the density, which is known to enforce isentropic behavior."

Reviewer 2: So it's artificially compressible SPH. The compressibility is artificial. When compressibility becomes an adjective, artificial becomes an adverb. Artificial compressible SPH would mean that the SPH is artificial and not the compressibility.

Author: Okay, right! "compressible SPH" doesn't make sense.
But "Entropically damped artiﬁcial compressibility for SPH" does make sense.

* *Structure Systems*

>     + Total lagrangian SPH (TLSPH): System to simulate elastic structure where all quantities are calculated with respect to the initial configuration [@O_Connor:2021].

Reviewer 2: ```suggestion
    + Total lagrangian SPH (TLSPH): Method to simulate elastic structures where all quantities are calculated with respect to the initial configuration [@O_Connor:2021].
```
Or "Scheme"

>     + Discrete element system: Discretization of a material system into a finite set of distinct, interacting mass elements [@Bicanic:2004], [@Cundall:1979].

Reviewer 2: ```suggestion
    + Discrete element method (DEM): Discretization of a material system into a finite set of distinct, interacting mass elements [@Bicanic:2004], [@Cundall:1979].
```
Maybe "granular material system" or so? Or can we avoid the word "system" here because we also use it for the data structure?

* *Boundary Systems*
    + Boundary system with several boundary models, where each model follows a different interaction rule.

>     + Open boundary system to simulate non-reflecting (open) boundary conditions [@Lastiwka:2009]

Reviewer 2: ```suggestion
    + Open boundary system to simulate non-reflecting (open) boundary conditions [@Lastiwka:2009].
```

* *Correction methods and models*
  + Density diffusion [@Antuono:2010]
  + Transport-velocity formulation (TVF) [@Adami:2013]
  + Intra-particle-force surface tension [@Akinci:2013]

* *Performance and parallelization*
  + Shared memory parallelism using multithreading
  + Highly optimized neighborhood search providing various approaches
  + GPU support

Validation can be performed by, for example, quantitatively comparing results using a post-process callback.

> \autoref{fig:beam_y_deflection} shows simulation results of TrixiParticles.jl (on the left) and [@O_Connor:2021] (on the right) compared against a reference value of [@Turek:2007].

Reviewer 2: ```suggestion
Validation can be performed, for example, by quantitatively comparing results using a post-process callback.
\autoref{fig:beam_y_deflection} shows simulation results of TrixiParticles.jl and [@O_Connor:2021] compared against a reference value of [@Turek:2007].
```

The object `Semidiscretization` couples the systems of a simulation and also creates the corresponding neighborhood searches for each system. The resulting ODE problem is then fed into the time integrator and is solved by an appropriate numerical time integration scheme.

![Main building blocks of TrixiParticles.jl \label{fig:structure}](structure.png){width=100%}

The code presented here can be found on [GitHub](https://github.com/trixi-framework/TrixiParticles.jl),  along with a detailed [manual](https://trixi-framework.github.io/TrixiParticles.jl/stable/) explaining how to use the package. Additionally, we provide tutorials explaining how to set up a simulation for fluids, structure mechanics or a fluid-structure interaction.
A collection of predefined simulation files can be found in the `examples` directory.

### Feature highlights

So far, the following feature highlights have been implemented:

* *Fluid Systems*
    + `WeaklyCmpressibleSPHSystem` (WCSPH): Standard SPH method originally developed by [@Monaghan:1977] to simulate astrophysics applications.

>     + `EntropicallyDampedSPHSystem` (EDAC): As opposed to the WCSPH scheme, which uses an equation of state, this scheme uses a pressure evolution equation to calculate the pressure which is derived by [@Clausen:2013] and adapted to SPH by [@Ramachandran:2019].

Reviewer 2: ```suggestion
    + `EntropicallyDampedSPHSystem` (EDAC): As opposed to the WCSPH scheme, which uses an equation of state, this scheme uses a pressure evolution equation to calculate the pressure, which is derived by [@Clausen:2013] and adapted to SPH by [@Ramachandran:2019].
```

* *Structure Systems*

>     + `TotalLagrangianSPHSystem` (TLSPH): System to simulate elastic structure where the interaction is defined with stress tensors [@O_Connor:2021].

Reviewer 2: ```suggestion
    + `TotalLagrangianSPHSystem` (TLSPH): System to simulate elastic structure where all quantities are calculated with respect to the initial configuration [@O_Connor:2021].
```

>     + `DEMSystem`: Discrete element system [@Bicanic:2004], [@Cundall:1979].

Reviewer 2: Should we elaborate a little more here?

Author: word count (~1100) says no :D

Reviewer 2: Maybe just "Discrete element system for granular flows" or so?

* *Boundary Systems*
    + `BoundarySPHSystem` with several boundary models where each model follows a different interaction rule.
    + `OpenBoundarySPHSystem`: System to simulate non-reflecting (open) boundary conditions [@Lastiwka:2009]

* *Correction methods and models*
  + Density diffusion [@Antuono:2010]
  + Transport-velocity formulation (TVF) [@Adami:2013]
  + Intra-particle-force surface tension [@Akinci:2013]

>   + Kernel corrections [@Akinci:2013], [@Bonet:1999], [@Basa:2009]

Reviewer 2: We're not actively using these, as it doesn't yield the results that we would like to see. I suggest we don't list them here.

Reviewer 2: Yes.

* *Performance*

>   + Efficient grid neighborhood search

Reviewer 2: I wouldn't list the NHS here. It's necessary to have NHS, and we didn't implement anything new here.

  + GPU support

\autoref{fig:falling_sphere} illustrates an example of our simulation results. In this example, an elastic sphere modeled with TLSPH falls into a tank filled with water, modeled by WCSPH.

![Elastic sphere falling into a tank filled with water. \label{fig:falling_sphere}](falling_sphere_combined_nonstick_4k_178.png){width=100%}

Validation can be performed by, for example, quantitatively comparing results using a post-process callback.
\autoref{fig:beam_y_deflection} shows simulation results of TrixiParticles.jl (on the left) and [@O_Connor:2021] (on the right) compared against a reference value of [@Turek:2007].

> The curves show the y-deflection of the tip of a beam oscillating under its own weight.

Reviewer 2: ```suggestion
The plots show the y-deflection of the tip of a beam oscillating under its own weight.
```

# Summary

TrixiParticles.jl, part of the Trixi Framework [@schlottkelakemper2020trixi], is an innovative Julia-based open-source software designed for particle-based multiphysics simulations.
It aims to handle complex geometries and specialized applications, such as computational fluid dynamics (CFD) and structural dynamics,
by offering a versatile platform for particle-based methods.
To increase flexibility, TrixiParticles.jl facilitates the easy addition of new particle systems and their interactions.
Coupling of different particle systems is very user friendly, thus multiphysics simulations such as fluid-structure interaction (FSI) can be set up very quickly.
This is further complemented by allowing simulations to be configured directly with Julia code.
This feature not only simplifies the integration of custom functionalities but also promotes rapid prototyping,

> establishing TrixiParticles.jl as a versatile and adaptable tool for advancing multiphysics simulations in various scientific and engineering fields.

Reviewer 2: I like this new summary! Only this half sentence sounds like too much advertising.

Here, we give a brief overview of the software package TrixiParticles.jl, starting with a description of the scientific background before going on to describe the functionality and benefit in more detail.
Finally, a few exemplary results and implemented features are briefly presented.

# Statement of need

Numerical simulations, such as CFD, structural mechanics, thermodynamics, or magnetohydrodynamics pose several challenges, especially when simulating real-world problems.
These simulations involve challenges such as dealing with complex geometries, free surfaces,
deformable boundaries and moving material interfaces as well as the coupling of different physics disciplines.

One way to ease these challenges is to use particle-based methods.
These methods offer two different approaches: the particles are either treated as physical particles or as mathematical interpolation points.
The former case refers to methods that model separate, discrete particles with rotational degrees of freedom such as the Discrete Element Method (DEM) proposed by [@Cundall:1979].

> Whereas the latter case refers to the smoothed particle hydrodynamics (SPH) which is a numerical discretization method for solving continuum mechanic problems.

Reviewer 2: ```suggestion
Whereas the latter case refers to methods like the smoothed particle hydrodynamics (SPH), which is a numerical discretization method for solving continuum mechanic problems.
```

Reviewer: ```suggestion
Whereas the latter case refers to methods such as Smoothed Particle Hydrodynamics (SPH), which is a numerical discretization method for solving problems in continuum mechanics.
```

In TrixiParticles.jl, particles of a single particle-based method are grouped into what we refer to as a \emph{system}.
The interaction between two particles is defined by the types of their systems. This approach makes it easy to add new methods and different physics.

> ![Particles of two different systems in a simulation domain. The black and gray dashed circles represent the search radii for neighboring particles of particles $a$ and $b$, respectively.\label{fig:systems}](systems.png){width=40%}

Reviewer 2: ```suggestion
![Particles of two different systems in a simulation domain. The black and gray dashed circles represent the search radii for neighbors of particles $a$ and $b$, respectively.\label{fig:systems}](systems.png){width=40%}
```

\autoref{fig:structure} depicts the basic building blocks of TrixiParticles.jl. A simulation essentially consists of spatial discretization (left block) and time integration (center block). For the latter, the Julia package [OrdinaryDiffEq.jl](https://github.com/SciML/OrdinaryDiffEq.jl) is used. The callbacks (right block) provide additional functionality and communicate with the time integration method during the simulation.

> The semidiscretization couples the systems of a simulation and also manage the corresponding neighborhood searches for each system.

Reviewer 2: ```suggestion
The semidiscretization couples the systems of a simulation and also manages the corresponding neighborhood searches for each system.
```
